### ISSUE TITLE
SNOW-684693: Small and large date/times not passed properly

### ISSUE THREAD

**Issue by Thomasdezeeuw** [snowflakedb/gosnowflake#678]
### Issue description

According to the documentation (https://docs.snowflake.com/en/sql-reference/data-types-datetime.html#data-types) Snowflake support years in the range 1582-9999, however when you pass small or large date times it will return incorrect values.

For example:

Value | Unix | Got
------|------|-----
9000-10-10 10:10:10 | 508864525485380608 | 1986-02-15 15:15:25.485380608
1600-10-10 10:10:10 | 6795135883709551616 |  2185-04-30 09:44:43.709551616

### Example code

Simply use `SELECT ?;` as query, passing one of the datetime mentioned above as `DATETIME`/`TIMESTAMP_NTZ`.

### Error log

N/A.

### Configuration

*Driver version:* v1.6.13.

*Go version:* go version go1.19.2 darwin/arm64.

*Server version:* 6.34.0.

*Client OS:* macOS.

**Comment by Thomasdezeeuw** [comment]
Any updates on this?

**Comment by Thomasdezeeuw** [comment]
I'm afraid I can't share the full program, but I think the following should reproduce it.

```go
	// Setup connection etc.
	var conn driver.QueryerContext

        timeValue, _ := time.Parse("2006-01-02 15:04:05", "1000-01-01 00:00:00")
	parameters := []driver.NamedValue {
		{ Ordinal: 1, Value: snowflake.DataTypeTimestampNtz},
		{ Ordinal: 2, Value: timeValue},
	};
	rows, err := conn.QueryContext(ctx, "SELECT $1", parameters)

	// Expected response in row[0][0] is "1000-01-01 00:00:00" as time.Time.
	// However we get "2169-02-08 23:09:07.419103232".
```

**Comment by sfc-gh-dszmolka** [comment]
maybe so, but currently i was unable to; getting
```
2023/03/30 14:52:14 failed to run a query. SELECT $1, err: sql: converting argument $1 type: unsupported type []driver.NamedValue, a slice of struct
exit status 1
```

instead. using a slightly modified version of [selectmany](https://github.com/snowflakedb/gosnowflake/blob/master/cmd/selectmany/selectmany.go) example program, this:
```go
// Example: Fetch many rows and allow cancel the query by Ctrl+C.
package main

import (
	"context"
	"database/sql"
        "database/sql/driver"
	"flag"
	"fmt"
	"log"
	"os"
	"os/signal"
	"strconv"
        "time"

	sf "github.com/snowflakedb/gosnowflake"
)

// getDSN constructs a DSN based on the test connection parameters
func getDSN() (string, *sf.Config, error) {
	env := func(k string, failOnMissing bool) string {
		if value := os.Getenv(k); value != "" {
			return value
		}
		if failOnMissing {
			log.Fatalf("%v environment variable is not set.", k)
		}
		return ""
	}

	account := env("SNOWFLAKE_TEST_ACCOUNT", true)
	user := env("SNOWFLAKE_TEST_USER", true)
	password := env("SNOWFLAKE_TEST_PASSWORD", true)
	host := env("SNOWFLAKE_TEST_HOST", false)
	port := env("SNOWFLAKE_TEST_PORT", false)
	protocol := env("SNOWFLAKE_TEST_PROTOCOL", false)

	portStr, err := strconv.Atoi(port)
	if err != nil {
		return "", nil, err
	}
	cfg := &sf.Config{
		Account:  account,
		User:     user,
		Password: password,
		Host:     host,
		Port:     portStr,
		Protocol: protocol,
		Tracing: "debug",
	}

	dsn, err := sf.DSN(cfg)
	return dsn, cfg, err
}

// run is an actual main
func run(dsn string) {
	// handler interrupt signal
	ctx, cancel := context.WithCancel(context.Background())
	c := make(chan os.Signal, 1)
	defer close(c)
	signal.Notify(c, os.Interrupt)
	defer func() {
		signal.Stop(c)
	}()
	go func() {
		select {
		case <-c:
			cancel()
		case <-ctx.Done():
		}
	}()

	db, err := sql.Open("snowflake", dsn)
	if err != nil {
		log.Fatalf("failed to connect. err: %v", err)
	}
	defer db.Close()
	fmt.Printf("Executing a query. It may take long. You may stop by Ctrl+C.\n")
	timeValue, _ := time.Parse("2006-01-02 15:04:05", "1000-01-01 00:00:00")
	parameters := []driver.NamedValue {
		{ Ordinal: 1, Value: sf.DataTypeTimestampNtz},
		{ Ordinal: 2, Value: timeValue},
	};
	query := "SELECT $1"
 	rows, err := db.QueryContext(ctx, query, parameters)
	if err != nil {
		log.Fatalf("failed to run a query. %v, err: %v", query, err)
	}
	defer rows.Close()
	var v1 string
	fmt.Printf("Fetching the results. It may take long. You may stop by Ctrl+C.\n")
	counter := 0
	for rows.Next() {
		if err = rows.Scan(&v1); err != nil {
			log.Fatalf("failed to get result. err: %v", err)
		}
		if counter%10000 == 0 {
			fmt.Printf("data: %v\n", v1)
		}
		if counter%1000000 == 0 {
			//debug.FreeOSMemory()
		}
		counter++
	}
	if rows.Err() != nil {
		fmt.Printf("ERROR: %v\n", rows.Err())
		return
	}
	fmt.Printf("Congrats! You have successfully run %v with Snowflake DB!\n", query)
}

func main() {
	if !flag.Parsed() {
		flag.Parse()
	}

	dsn, cfg, err := getDSN()
	if err != nil {
		log.Fatalf("failed to create DSN from Config: %v, err: %v", cfg, err)
	}

	run(dsn)

```
i understand it's not possible to share the original full program which leads to the issue for you; but perhaps sharing a minimal runnable program which exhibits the issue could speed up the debug process. thank you in advance !

**Comment by Thomasdezeeuw** [comment]
> maybe so, but currently i was unable to; getting
> 
> ```
> 2023/03/30 14:52:14 failed to run a query. SELECT $1, err: sql: converting argument $1 type: unsupported type []driver.NamedValue, a slice of struct
> exit status 1
> ```

If you look at my example you'll see

```go
var conn driver.QueryerContext
```

In other words our connection is of type [`driver.QueryerContext`](https://pkg.go.dev/database/sql/driver#QueryerContext), which does accept a slice of `driver.NamedValue`.

> instead. using a slightly modified version of [selectmany](https://github.com/snowflakedb/gosnowflake/blob/master/cmd/selectmany/selectmany.go) example program, this:

I've merged our two examples:

```go
package main

import (
	"context"
	"database/sql/driver"
	"fmt"
	"io"
	"log"
	"os"
	"time"

	sf "github.com/snowflakedb/gosnowflake"
)

func getConf() (*sf.Config, error) {
	env := func(k string, failOnMissing bool) string {
		if value := os.Getenv(k); value != "" {
			return value
		}
		if failOnMissing {
			log.Fatalf("%v environment variable is not set.", k)
		}
		return ""
	}

	account := env("SNOWFLAKE_TEST_ACCOUNT", true)
	user := env("SNOWFLAKE_TEST_USER", true)
	password := env("SNOWFLAKE_TEST_PASSWORD", true)

	cfg := &sf.Config{
		Account:  account,
		User:     user,
		Password: password,
		Protocol: "https",
		Host:     account + ".snowflakecomputing.com",
		Port:     443,
		//Tracing:  "debug",
		Params: map[string]*string{},
	}
	return cfg, nil
}

func main() {
	cfg, err := getConf()
	if err != nil {
		log.Fatalf("failed to create Config: %v, err: %v", cfg, err)
	}

	ctx := context.Background()
	ctx = sf.WithHigherPrecision(ctx)
	sfDriver := sf.SnowflakeDriver{}
	c, err := sfDriver.OpenWithConfig(ctx, *cfg)
	if err != nil {
		log.Fatalf("failed to create connection, config: %v, err: %v", cfg, err)
	}
	defer c.Close()
	conn := c.(driver.QueryerContext)

	timeValue, err := time.Parse("2006-01-02 15:04:05", "1000-01-01 00:00:00")
	if err != nil {
		log.Fatalf("failed to parse time: %v", err)
	}
	parameters := []driver.NamedValue{
		{Ordinal: 1, Value: sf.DataTypeTimestampNtz},
		{Ordinal: 2, Value: timeValue},
	}

	fmt.Printf("parameters: %v\n", parameters)
	rows, err := conn.QueryContext(ctx, "SELECT ?", parameters)
	if err != nil {
		log.Fatalf("failed to run query: %v", err)
	}
	defer rows.Close()

	fmt.Printf("columns: %v\n", rows.Columns())

	scanValues := make([]driver.Value, 1)
	for {
		if err := rows.Next(scanValues); err == io.EOF {
			break
		} else if err != nil {
			log.Fatalf("failed to run query: %v", err)
		}

		fmt.Printf("values: %v\n", scanValues)
	}

	// Expected response in row[0][0] is "1000-01-01 00:00:00" as time.Time.
	// However we get "2169-02-08 23:09:07.419103232".
}
```

For me this outputs the following:

```
parameters: [{ 1 [6]} { 2 1000-01-01 00:00:00 +0000 UTC}]
columns: [?]
values: [2169-02-08 23:09:07.419103232 +0000 UTC]
```

**Comment by sfc-gh-dszmolka** [comment]
this is super helpful, thank you ! not only managed to reproduce the issue with your repro script, but I have some idea what might be wrong here: the nanosecond Epoch representation for these timestamps seem to be out of bounds for the unsigned int64 datatype minimum and maximum value.

i.e. anything **earlier** than 21 September 1677 00:12:43.145 UTC (epoch `-9223372036854775808`, also `math.MinInt64`) or anything **later** than 11 April 2262 23:47:16.854 UTC (epoch `9223372036854775807`, also `math.MaxInt64`) is possibly affected, despite the Snowflake engine itself would support such small or such big dates.

we'll take a look and i'll update the progress here on this issue.

**Comment by sfc-gh-dszmolka** [comment]
#788

**Comment by sfc-gh-dszmolka** [comment]
above fix PR is merged and the fix will be part of the next release, expected in second half of May

**Comment by sfc-gh-dszmolka** [comment]
fix released with version [1.6.21](https://github.com/snowflakedb/gosnowflake/releases/tag/v1.6.21), closing this issue
